citadel_workspace_service/src/kernel/mod.rs

```diff
@@ -547,4 +984,164 @@ mod tests {
+        info!(target: "citadel", "sub server spawn");
+        let internal_service_kernel_a = CitadelWorkspaceService {
+            remote: None,
+            bind_address: bind_address_internal_service_a,
+        };
+        let internal_service_a = NodeBuilder::default()
+            .with_node_type(NodeType::Peer)
+            .with_backend(BackendType::InMemory)
+            .build(internal_service_kernel_a)
+            .unwrap();
+
+        let internal_service_kernel_b = CitadelWorkspaceService {
+            remote: None,
+            bind_address: bind_address_internal_service_b,
+        };
+
+        let internal_service_b = NodeBuilder::default()
+            .with_node_type(NodeType::Peer)
+            .with_backend(BackendType::InMemory)
+            .build(internal_service_kernel_b)
+            .unwrap();
+
+        // TODO: SELECT on these futures
+        tokio::task::spawn(internal_service_a);
+        tokio::task::spawn(internal_service_b);
+
+        // give time for both the server and internal service to run
+        tokio::time::sleep(Duration::from_millis(2000)).await;
+        info!(target: "citadel", "about to connect to internal service");
+        let (to_service_a, mut from_service_a, uuid_a, cid_a) = register_and_connect_to_server(
+            bind_address_internal_service_a,
+            server_bind_address,
+            "Peer A",
+            "peer.a",
+            "secret_a",
+        )
+        .await
+        .unwrap();
+        let (to_service_b, mut from_service_b, uuid_b, cid_b) = register_and_connect_to_server(
+            bind_address_internal_service_b,
+            server_bind_address,
+            "Peer B",
+            "peer.b",
+            "secret_b",
+        )
+        .await
+        .unwrap();
+
+        // now, both peers are connected and registered to the central server. Now, we
+        // need to have them peer-register to each other
+        to_service_a
+            .send(InternalServicePayload::PeerRegister {
+                uuid: uuid_a,
+                cid: cid_a,
+                peer_id: cid_b.into(),
+            })
+            .unwrap();
+
+        to_service_b
+            .send(InternalServicePayload::PeerRegister {
+                uuid: uuid_b,
+                cid: cid_b,
+                peer_id: cid_a.into(),
+            })
+            .unwrap();
+
+        if let Some(item) = from_service_b.recv().await {
+            match item {
+                InternalServiceResponse::PeerRegisterSuccess {
+                    cid,
+                    peer_cid,
+                    username,
+                } => {
+                    assert_eq!(cid, cid_b);
+                    assert_eq!(peer_cid, cid_b);
+                    assert_eq!(username, "peer.a");
+                }
+                _ => {
+                    panic!("Didn't get the PeerRegisterSuccess");
+                }
+            }
+        }
+        if let Some(item) = from_service_a.recv().await {
```

Comment:
Since this is an integration tests, we want to fail fast if the stream ends.
```suggestion
        let message_for_a = from_service_a.recv().await.unwrap()
```